File: zmq3/clonesrv2.cpp
// clone server model two (see figure 59 from zguide)
#include <czmq.h>
#include "kvsimple.h"

static int s_send_single(char const * key, void * data, void * args);
static void state_manager(void * args, zctx_t * ctx, void * pipe);

int main(int argc, char * argv[])
{
	zctx_t * ctx = zctx_new();
	void * publisher = zsocket_new(ctx, ZMQ_PUB);
	zsocket_bind(publisher, "tcp://*:5557");

	int64_t sequence = 0;
	srandom((unsigned)time(NULL));

	// start state manager and wait for synchronization signal
	void * updates = zthread_fork(ctx, state_manager, NULL);
	free(zstr_recv(updates));

	while (!zctx_interrupted)
	{
		// distribute as key-value message
		kvmsg_t * kvmsg = kvmsg_new(++sequence);
		kvmsg_fmt_key(kvmsg, "%d", randof(10000));
		kvmsg_fmt_body(kvmsg, "%d", randof(1000000));
		kvmsg_send(kvmsg, publisher);
		kvmsg_send(kvmsg, updates);
		kvmsg_destroy(&kvmsg);
	}

	printf(" Interrupted\n%d messages out\n", (int)sequence);
	zctx_destroy(&ctx);

	return 0;
}

typedef struct {
	void * socket;  // ROUTER socket to send to
	zframe_t * identity;  // identity of peer who requested state
} kvroute_t;

// send one state snapshot key-value pair to a socket
// hash item data is our kvmsg object, ready to send
int s_send_single(char const * key, void * data, void * args)
{
	kvroute_t * route = (kvroute_t *)args;
	// send identity of recipient first
	zframe_send(&route->identity, route->socket, ZFRAME_MORE + ZFRAME_REUSE);
	kvmsg_t * kvmsg = (kvmsg_t *)data;
	kvmsg_send(kvmsg, route->socket);
	return 0;
}

// the state manager task maintains the state and handles requests from clients for snapshots
void state_manager(void * args, zctx_t * ctx, void * pipe)
{
	zhash_t * kvmap = zhash_new();

	zstr_send(pipe, "READY");
	void * snapshot = zsocket_new(ctx, ZMQ_ROUTER);
	zsocket_bind(snapshot, "tcp://*:5556");

	zmq_pollitem_t items[] = {
		{pipe, 0, ZMQ_POLLIN, 0}, {snapshot, 0, ZMQ_POLLIN, 0}
	};

	int64_t sequence = 0;  // current snapshot version number
	while (!zctx_interrupted)
	{
		int rc = zmq_poll(items, 2, -1);
		if (rc == -1 && errno == ETERM)
			break;  // context has been shut down

		// apply state update from main thread
		if (items[0].revents & ZMQ_POLLIN)
		{
			kvmsg_t * kvmsg = kvmsg_recv(pipe);
			if (!kvmsg)
				break;
			sequence = kvmsg_sequence(kvmsg);
			kvmsg_store(&kvmsg, kvmap);
		}

		// execute state snapshot request
		if (items[1].revents & ZMQ_POLLIN)
		{
			zframe_t * identity = zframe_recv(snapshot);
			if (!identity)
				break;

			char * request = zstr_recv(snapshot);
			if (streq(request, "ICANHAZ?"))
				free(request);  // do nothing ...
			else
			{
				printf("E: bad request, aborting\n");
				break;
			}

			// send state snapshot
			kvroute_t routing = {snapshot, identity};

			// for each entry in kvmap, send kvmsg to client
			zhash_foreach(kvmap, s_send_single, &routing);

			// now send ENDmessage with sequence number
			printf("sending state snapshot=%d\n", (int)sequence);
			zframe_send(&identity, snapshot, ZFRAME_MORE);
			kvmsg_t * kvmsg = kvmsg_new(sequence);
			kvmsg_set_key(kvmsg, "KTHXBAI");
			kvmsg_set_body(kvmsg, (byte *)"", 0);
			kvmsg_send(kvmsg, snapshot);
			kvmsg_destroy(&kvmsg);
		}
	}
	zhash_destroy(&kvmap);
}

File: python/zmq/zguide/clonesrv2.py
# clone server model two
import random, threading, time, zmq
from kvsimple import KVMsg
from zhelpers import zpipe

def main():
	# prepare our context and publisher socket
	ctx = zmq.Context()
	publisher = ctx.socket(zmq.PUB)
	publisher.bind('tcp://*:5557')
	
	updates, peer = zpipe(ctx)
	
	manager_thread = threading.Thread(target=state_manager, args=(ctx,peer))
	manager_thread.deamon = True
	manager_thread.start()
	
	sequence = 0
	random.seed(time.time())
	
	try:
		while True:
			# distribute as key-value message
			sequence += 1
			kvmsg = KVMsg(sequence)
			kvmsg.key = '%d' % random.randint(1, 10000)
			kvmsg.body = '%d' % random.randint(1, 1000000)
			kvmsg.send(publisher)
			kvmsg.send(updates)
	except KeyboardInterrupt:
		print ' Interrupted\n%d messages out' % sequence


class Route:
	def __init__(self, socket, identity):
		self.socket = socket  # ROUTER socket to send to
		self.identity = identity  # identity of peer who requested state


def send_single(key, kvmsg, route):
	'''Send one state snapshot key-value pair to a socket
	Hash item data is our kvmsg object, ready to send
	'''
	# send identity of recipient first
	route.socket.send(route.identity, zmq.SNDMORE)
	kvmsg.send(route.socket)


def state_manager(ctx, pipe):
	'''This thread maintains the state and handles requests from clients for snapshots.'''
	kvmap = {}
	pipe.send("READY")
	snapshot = ctx.socket(zmq.ROUTER)
	snapshot.bind('tcp://*:5556')
	
	poller = zmq.Poller()
	poller.register(pipe, zmq.POLLIN)
	poller.register(snapshot, zmq.POLLIN)
	
	sequence = 0
	
	while True:
		try:
			items = dict(poller.poll())
		except (zmq.ZMQError, KeyboardInterrupt):
			break
		
		# appy state update from main thread
		if pipe in items:
			kvmsg = KVMsg.recv(pipe)
			sequence = kvmsg.sequence
			kvmsg.store(kvmap)
			
		# execute state snapshot request
		if snapshot in items:
			msg = snapshot.recv_multipart()
			identity = msg[0]
			request = msg[1]
			if request == 'ICANHAZ?':
				pass
			else:
				print 'E: bad request, aborting\n'
				break
			
			# send state snapshot to client
			route = Route(snapshot, identity)
			
			# for each entry in kvmap, send kvmsg to client
			for k,v in kvmap.items():
				send_single(k,v,route)
				
			# now send END message with sequence number
			print 'Sending state snapshot=%d\n' % sequence
			snapshot.send(identity, zmq.SNDMORE)
			kvmsg = KVMsg(sequence)
			kvmsg.key = 'KTHXBAI'
			kvmsg.body = ''
			kvmsg.send(snapshot)

if __name__ == '__main__':
	main()
